calyx-backend/src/firrtl.rs

```diff
@@ -34,14 +35,35 @@ impl Backend for FirrtlBackend {
 
     fn emit(ctx: &ir::Context, file: &mut OutputFile) -> CalyxResult<()> {
         let out = &mut file.get_write();
-        let mut top_level_component = String::from("main");
-        // Quick pass to check whether there exists a top-level component that we should replace main with.
+        writeln!(out, "circuit {}:", ctx.entrypoint)?;
+        // Pass to output any necessary extmodule statements (for primitive calls)
+        let mut extmodule_set: HashSet<String> = HashSet::new();
         for comp in ctx.components.iter() {
-            if comp.attributes.has(ir::BoolAttr::TopLevel) {
-                top_level_component = comp.name.to_string().clone();
+            for cell in comp.cells.iter() {
+                let cell_borrowed = cell.as_ref().borrow();
+                if cell_borrowed.type_name().is_some() {
+                    if let ir::CellType::Primitive {
+                        name,
+                        param_binding,
+                        is_comb: _,
+                        latency: _,
+                    } = &cell_borrowed.prototype
+                    {
+                        let curr_module_name =
+                            get_primitive_module_name(name, param_binding);
+                        if !extmodule_set.contains(&curr_module_name) {
+                            emit_primitive_extmodule(
+                                &curr_module_name,
+                                name,
+                                param_binding,
+                                out,
+                            )?;
+                            extmodule_set.insert(curr_module_name);
+                        }
+                    };
+                }
             }
         }
-        writeln!(out, "circuit {}:", top_level_component)?;
         for comp in ctx.components.iter() {
```

Comment:
if desired this could be
```
for comp in &ctx.components
```
since `&Vec<_>` implements `IntoIter`.

Either way works and I don't think there's a best practice here but I figured that I'd mention the option